pkg/declarative/v2/reconciler.go

```diff
@@ -0,0 +1,382 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func NewFromManager(mgr manager.Manager, prototype Object, options ...Option) reconcile.Reconciler {
+	r := &Reconciler{}
+	r.prototype = prototype
+	r.Options = DefaultOptions().Apply(WithManager(mgr)).Apply(options...)
+	return r
+}
+
+type Reconciler struct {
+	prototype Object
+	*Options
+}
+
+type ConditionType string
+
+const (
+	ConditionTypeResources    ConditionType = "Resources"
+	ConditionTypeInstallation ConditionType = "Installation"
+)
+
+type ConditionReason string
+
+const (
+	ConditionReasonResourcesAreAvailable ConditionReason = "ResourcesAvailable"
+	ConditionReasonReady                 ConditionReason = "Ready"
+)
+
+//nolint:funlen,nestif,gocognit,gocyclo,cyclop,maintidx //TODO discuss if worth breaking up or if more readable as is.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	obj := r.prototype.DeepCopyObject().(Object)
+
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
+		logger.Info(req.NamespacedName.String() + " got deleted!")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if r.ShouldSkip(ctx, obj) {
+		return ctrl.Result{}, nil
+	}
+
+	// ServerSideApply requires nil ManagedFields
+	obj.SetManagedFields(nil)
+
+	status := obj.GetStatus()
+
+	if !obj.GetDeletionTimestamp().IsZero() && obj.GetStatus().State != StateDeleting {
+		obj.SetStatus(status.WithState(StateDeleting).WithOperation("switched to deleting"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	// add finalizer
+	if controllerutil.AddFinalizer(obj, r.Finalizer) {
+		return r.ssa(ctx, obj)
+	}
+
+	var (
+		resourceCondition = metav1.Condition{
+			Type:               string(ConditionTypeResources),
+			Reason:             string(ConditionReasonResourcesAreAvailable),
+			Status:             metav1.ConditionFalse,
+			Message:            "resources are parsed and ready for use",
+			ObservedGeneration: obj.GetGeneration(),
+		}
+		installationCondition = metav1.Condition{
+			Type:               string(ConditionTypeInstallation),
+			Reason:             string(ConditionReasonReady),
+			Status:             metav1.ConditionFalse,
+			Message:            "installation is ready and resources can be used",
+			ObservedGeneration: obj.GetGeneration(),
+		}
+	)
+
+	// Processing
+	if status.State == "" {
+		presetConditions := 2
+		status.Conditions = make([]metav1.Condition, 0, presetConditions)
+		meta.SetStatusCondition(&status.Conditions, resourceCondition)
+		meta.SetStatusCondition(&status.Conditions, installationCondition)
+		status.Synced = make([]Resource, 0)
+		obj.SetStatus(status.WithState(StateProcessing).WithOperation("switched to processing"))
+		return r.ssaStatus(ctx, obj)
+	}
+
+	spec, err := r.Spec(ctx, obj)
```

Comment:
By far, this Spec() implementation doesn't return actual err, only init Spec struct, what's the concern here for this error handling?

Reply:
this will later on error out for example if the OCI image resolve in module manager will fail. Then the translation of OCI layers in Manifest to the Spec will fail and thus the Module Manager needs a point to return an error. Thus this err needs to be propagated and cause a Failure/Error on the Module/Manifest CR

Reply:
Do you mean we will have more `SpecResolver` implementation later?

Reply:
There will be at least one for ModuleManager to resolve a Manifest => v2.Spec. This process can surely error if the manifest contains invalid data for any reason.

Reply:
I am assuming if there will be more versions of Manifest (v1alpha1, v1alpha2....) it would result it multiple resolvers